src/devices/stepper.rs

```diff
@@ -0,0 +1,526 @@
+// phidget-rs/src/stepper.rs
+//
+// Copyright (c) 2023, Frank Pagliughi
+// implemented by willmendil 2024.
+//
+// This file is part of the 'phidget-rs' library.
+//
+//
+// Licensed under the MIT license:
+//   <LICENSE or http://opensource.org/licenses/MIT>
+// This file may not be copied, modified, or distributed except according
+// to those terms.
+//
+
+use crate::{AttachCallback, DetachCallback, GenericPhidget, Phidget, Result, ReturnCode};
+use phidget_sys::{self as ffi, PhidgetHandle, PhidgetStepperHandle as StepperHandle};
+use std::{mem, os::raw::c_void, ptr};
+
+/// The function type for the safe Rust position change callback.
+pub type PositionChangeCallback = dyn Fn(&Stepper, f64) + Send + 'static;
+/// The function type for the safe Rust velocity change callback.
+pub type VelocityChangeCallback = dyn Fn(&Stepper, f64) + Send + 'static;
+/// The function type for the safe Rust stop callback.
+pub type StoppedCallback = dyn Fn(&Stepper) + Send + 'static;
+
+/// Phidget Stepper sensor
+pub struct Stepper {
+    // Handle to the sensor for the phidget22 library
+    chan: StepperHandle,
+    // Double-boxed StepperCallback, if registered
+    cb: Option<*mut c_void>,
+    // Double-boxed attach callback, if registered
+    attach_cb: Option<*mut c_void>,
+    // Double-boxed detach callback, if registered
+    detach_cb: Option<*mut c_void>,
+}
+
+/// ControlMode for stepper
+#[derive(Copy, Clone)]
+pub enum ControlMode {
+    /// Step: Control the motor by setting a target position.
+    Step,
+    /// Run: Control the motor by selecting a target velocity (sign indicates direction). The motor will rotate continuously in the chosen direction.
+    Run,
+}
```

Comment:
Just a small thing,.. If you're making enums that will convert to integers for the FFI layer, it's a little easier to specify the integer size you want, like `#[repr(u32)]` (or `u8`, `i32`, etc) and the numeric values for each discriminant. And, for a unit-only enumeration like this, unless there's a specific reason not to, just go ahead and derive comparators and hashing capabilities. All together:

```
#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[repr(u32)]
pub enum ControlMode {
    /// Step: Control the motor by setting a target position.
    Step = 0,
    /// Run: Control the motor by selecting a target velocity (sign indicates direction). The motor will rotate continuously in the chosen direction.
    Run = 1,
}
```

Then, later you can convert to u32 by a simple cast:
```
    pub fn set_control_mode(&self, control_mode: ControlMode) -> Result<()> {
        ReturnCode::result(unsafe { ffi::PhidgetStepper_setControlMode(self.chan, control_mode as c_uint) })?;
        Ok(())
    }
```